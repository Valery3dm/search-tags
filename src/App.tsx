import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';

import { ItemsState } from './store/types/item';
import { LastSearch, PreLoader, Cards, SearchPanel } from './components';
import { useActions } from './hooks/useAction';

import { AppStyled } from './styled';

const App: React.FC = () => {
 
  const {
    itemsList,
    isLoaded,
    listOfThreeLastItems
  } = useSelector((state: ItemsState) => state);

  const { fetchItemsAction, setPage } = useActions();

  let { page } = useSelector((state: ItemsState) => state);

  const handlerPage = () => new Promise(res => res(setPage(++page)));
  const handleFetch = () => new Promise(res => res(fetchItemsAction()))

  useEffect(() => {
    const scrollHandler = async (e: any) => {
      if(e.target.documentElement.scrollHeight - (e.target.documentElement.scrollTop + window.innerHeight) < 100) {
        await handlerPage();
        await handleFetch();
      }
    }
    document.addEventListener('scroll', scrollHandler);
    return function() {
      document.removeEventListener('scroll', scrollHandler);
    }
  },[]);

  

  return (
    <AppStyled>
      <SearchPanel />
      <LastSearch listOfThreeLastItems={listOfThreeLastItems} />
      {isLoaded ? (
        <Cards itemsList={itemsList}/>
      ) : (
        <PreLoader/>
      )}
    </AppStyled>
  );
};

export default App;
