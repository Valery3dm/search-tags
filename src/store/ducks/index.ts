import { combineReducers } from 'redux';
import { itemReducer } from './itemReducer';

export const rootReducer = combineReducers({
    item: itemReducer
})

export type RootState = ReturnType<typeof rootReducer>
