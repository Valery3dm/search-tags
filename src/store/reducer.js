const initialState = {
  itemsList: [],
  listOfThreeLastItems: [],
  inputItem: ''
};

const FETCH_ITEMS = "FETCH_ITEMS";
const SET_THREE_LAST_ITEMS = "SET_THREE_LAST_ITEMS";
const SET_INPUT_ITEM = "SET_INPUT_ITEM";

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_ITEMS:
      return { ...state, itemsList: action.payload };
    case SET_THREE_LAST_ITEMS:
      return { ...state,
         listOfThreeLastItems: [ ...state.listOfThreeLastItems, action.payload]
        };
    case SET_INPUT_ITEM:
      return { ...state, inputItem: action.payload };
    default:
      return state;
  }
};

export default reducer;