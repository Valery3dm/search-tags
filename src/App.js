import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import LastSearch from "./components/last-search/last-search";
import SearchPanel from "./components/search-panel/search-panel";
import Cards from "./components/cards/cards";
import PreLoader from "./components/pre-loader/pre-loader";

import { duck } from "./store/ducks/widgets";

import "./App.css";

const App = () => {
  const dispatch = useDispatch();
  const itemsList = useSelector((state) => state.itemsList);
  const listOfThreeLastItems = useSelector(
    (state) => state.listOfThreeLastItems
  );

  useEffect(() => {
    dispatch(duck.actionCreators.fetchItemsAction());
  }, [dispatch]);

  return (
    <div className="app">
      <SearchPanel />
      <LastSearch listOfThreeLastItems={listOfThreeLastItems} />
      {itemsList ? <Cards itemsList={itemsList} /> : <PreLoader />}
    </div>
  );
};

export default App;
