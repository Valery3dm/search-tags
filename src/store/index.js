import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import createSagaMiddleware from "redux-saga";

import { duck } from "./ducks/widgets";


const sagaMiddleware = createSagaMiddleware();

const store = createStore(duck.reducers.reducer, composeWithDevTools(applyMiddleware(sagaMiddleware)));

export default store;

sagaMiddleware.run(duck.saga);
