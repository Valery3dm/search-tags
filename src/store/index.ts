import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import { sagaFetchItemsWatcher } from './saga/sagaFetchItems';
import { itemReducer } from './ducks/itemReducer';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  itemReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

export default store;

sagaMiddleware.run(sagaFetchItemsWatcher);