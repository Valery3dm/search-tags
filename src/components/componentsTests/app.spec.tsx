import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import store from '../../store';

import App from '../../App';
import { BrowserRouter } from 'react-router-dom';

Enzyme.configure({ adapter: new Adapter() });

describe('App component', () => {

    const component = mount(
        <Provider store={store}>
            <BrowserRouter>
                <App />
            </BrowserRouter> 
        </Provider>
    );

    it('App snapshots component', () => {
        expect(toJson(component.find(App))).toMatchSnapshot();
    });

});