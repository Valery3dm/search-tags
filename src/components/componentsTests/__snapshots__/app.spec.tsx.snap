// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshots component 1`] = `
<App>
  <styled.div>
    <div
      className="sc-bdnxRM iaMMfV"
    >
      <SearchPanel>
        <Styled(Component)>
          <ForwardRef(InputGroup)
            className="sc-fujyAs kTYqnT"
          >
            <div
              className="sc-fujyAs kTYqnT input-group"
            >
              <FormControl
                aria-describedby="basic-addon2"
                aria-label="Recipient's username"
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder="type your tag here"
                value=""
              >
                <input
                  aria-describedby="basic-addon2"
                  aria-label="Recipient's username"
                  className="form-control"
                  onChange={[Function]}
                  onKeyPress={[Function]}
                  placeholder="type your tag here"
                  value=""
                />
              </FormControl>
              <InputGroupAppend>
                <div
                  className="input-group-append"
                >
                  <Link
                    to="/"
                  >
                    <LinkAnchor
                      href="/"
                      navigate={[Function]}
                    >
                      <a
                        href="/"
                        onClick={[Function]}
                      >
                        <Button
                          active={false}
                          disabled={false}
                          onClick={[Function]}
                          variant="outline-secondary"
                        >
                          <button
                            className="btn btn-outline-secondary"
                            disabled={false}
                            onClick={[Function]}
                            type="button"
                          >
                            Search
                          </button>
                        </Button>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </InputGroupAppend>
            </div>
          </ForwardRef(InputGroup)>
        </Styled(Component)>
      </SearchPanel>
      <LastSearch
        listOfThreeLastItems={Array []}
      >
        <styled.div>
          <div
            className="sc-jSFjdj bcarFw"
          >
            <div>
              <styled.ul>
                <ul
                  className="sc-gKAaRy fgCheT"
                />
              </styled.ul>
            </div>
          </div>
        </styled.div>
      </LastSearch>
      <PreLoader>
        <div>
          <styled.div>
            <div
              className="sc-iCoGMd dnJlUA"
            >
              &lt; ðŸŒ¸ &gt;
            </div>
          </styled.div>
          <div>
            Please enter TAG
          </div>
        </div>
      </PreLoader>
    </div>
  </styled.div>
</App>
`;
